packages/typescript/src/internal/ts_project.bzl

```diff
@@ -0,0 +1,192 @@
+    fields = {
+        "tsconfigs": "depset of tsconfig.json files",
+    },
+)
+
+def _ts_project_impl(ctx):
+    arguments = ctx.actions.args()
+    arguments.add_all([
+        "-p",
+        ctx.file.tsconfig.short_path,
+        "--outDir",
+        "/".join([ctx.bin_dir.path, ctx.label.package]),
+    ])
+    if len(ctx.outputs.typings_outs) > 0:
+        arguments.add_all([
+            "--declarationDir",
+            "/".join([ctx.bin_dir.path, ctx.label.package]),
+        ])
+
+    deps_depsets = []
+    for dep in ctx.attr.deps:
+        if _TsConfigInfo in dep:
+            deps_depsets.append(dep[_TsConfigInfo].tsconfigs)
+        if NpmPackageInfo in dep:
+            # TODO: we could maybe filter these to be tsconfig.json or *.d.ts only
+            # we don't expect tsc wants to read any other files from npm packages.
+            deps_depsets.append(dep[NpmPackageInfo].sources)
+        elif DeclarationInfo in dep:
+            deps_depsets.append(dep[DeclarationInfo].transitive_declarations)
+
+    inputs = ctx.files.srcs + depset(transitive = deps_depsets).to_list() + [ctx.file.tsconfig]
+    if ctx.attr.extends:
+        inputs.extend(ctx.files.extends)
+    outputs = ctx.outputs.js_outs + ctx.outputs.map_outs + ctx.outputs.typings_outs + ctx.outputs.typing_maps_outs
+
+    if len(outputs) == 0:
+        return []
+
+    run_node(
+        ctx,
+        inputs = inputs,
+        arguments = [arguments],
+        outputs = outputs,
+        executable = "tsc",
+        progress_message = "Compiling TypeScript project %s" % ctx.file.tsconfig.short_path,
+    )
+
+    runtime_files = depset(ctx.outputs.js_outs + ctx.outputs.map_outs)
+    typings_files = ctx.outputs.typings_outs + [s for s in ctx.files.srcs if s.path.endswith(".d.ts")]
+
+    return [
+        DeclarationInfo(
+            declarations = depset(typings_files),
+            transitive_declarations = depset(typings_files, transitive = [
+                dep[DeclarationInfo].transitive_declarations
+                for dep in ctx.attr.deps
+            ]),
+        ),
+        DefaultInfo(
+            files = runtime_files,
+            runfiles = ctx.runfiles(
+                transitive_files = runtime_files,
+                collect_default = True,
+            ),
+        ),
+        _TsConfigInfo(tsconfigs = depset([ctx.file.tsconfig], transitive = [
+            dep[_TsConfigInfo].tsconfigs
+            for dep in ctx.attr.deps
+            if _TsConfigInfo in dep
+        ])),
+    ]
+
+ts_project = rule(
+    implementation = _ts_project_impl,
+    attrs = dict(_ATTRS, **_OUTPUTS),
+)
+
+def _out_paths(srcs, ext):
+    return [f[:f.rindex(".")] + ext for f in srcs if not f.endswith(".d.ts")]
+
+def ts_project_macro(
+        name = "tsconfig",
+        srcs = None,
+        deps = [],
+        extends = None,
+        declaration = True,
+        source_map = False,
+        declaration_map = False,
+        emit_declaration_only = False,
```

Comment:
maybe instead have a flag that says `emit_js = True`? then this can be derived from the inverse of that flag?

Reply:
I'm trying to keep these flags named identically to the tsconfig file so the mapping between the two is totally obvious.

Reply:
Ah right, I wasn't aware of a flag in the tsconfig of the same name, sounds good!